File: src/jsc/GarbageCollectionController.zig
//! Garbage Collection Controller for Bun's JavaScript runtime
//!
//! This controller intelligently schedules garbage collection to run at optimal times,
//! such as when HTTP requests complete, during idle periods, or when memory usage
//! has grown significantly since the last collection cycle.
//!
//! The controller works in conjunction with JavaScriptCore's built-in GC timers to
//! provide additional collection opportunities, particularly in scenarios where:
//! - JavaScript code is not actively executing (e.g., waiting for I/O)
//! - The event loop is idle but memory usage has increased
//! - Long-running operations have allocated significant memory
//!
//! Key features:
//! - Adaptive timing based on heap growth patterns
//! - Configurable intervals via BUN_GC_TIMER_INTERVAL environment variable
//! - Can be disabled via BUN_GC_TIMER_DISABLE for debugging/testing
//!
//! Thread Safety: This type must be unique per JavaScript thread and is not
//! thread-safe. Each VirtualMachine instance should have its own controller.

const GarbageCollectionController = @This();

gc_timer: *uws.Timer = undefined,
gc_last_heap_size: usize = 0,
gc_last_heap_size_on_repeating_timer: usize = 0,
heap_size_didnt_change_for_repeating_timer_ticks_count: u8 = 0,
#idle_full_gcs_fired: u8 = 0,
gc_timer_state: GCTimerState = GCTimerState.pending,
gc_repeating_timer: *uws.Timer = undefined,
gc_timer_interval: i32 = 0,
gc_repeating_timer_fast: bool = true,
disabled: bool = false,

pub fn init(this: *GarbageCollectionController, vm: *VirtualMachine) void {
    const actual = uws.Loop.get();
    this.gc_timer = uws.Timer.createFallthrough(actual, this);
    this.gc_repeating_timer = uws.Timer.createFallthrough(actual, this);
    actual.internal_loop_data.jsc_vm = vm.jsc_vm;

    if (comptime Environment.isDebug) {
        if (bun.env_var.BUN_TRACK_LAST_FN_NAME.get()) {
            vm.eventLoop().debug.track_last_fn_name = true;
        }
    }

    var gc_timer_interval: i32 = 1000;
    if (vm.transpiler.env.get("BUN_GC_TIMER_INTERVAL")) |timer| {
        if (std.fmt.parseInt(i32, timer, 10)) |parsed| {
            if (parsed > 0) {
                gc_timer_interval = parsed;
            }
        } else |_| {}
    }
    this.gc_timer_interval = gc_timer_interval;

    if (vm.transpiler.env.get("BUN_GC_RUNS_UNTIL_SKIP_RELEASE_ACCESS")) |val| {
        if (std.fmt.parseInt(c_int, val, 10)) |parsed| {
            if (parsed >= 0) {
                VirtualMachine.Bun__defaultRemainingRunsUntilSkipReleaseAccess = parsed;
            }
        } else |_| {}
    }

    this.disabled = vm.transpiler.env.has("BUN_GC_TIMER_DISABLE");

    if (!this.disabled)
        this.gc_repeating_timer.set(this, onGCRepeatingTimer, gc_timer_interval, gc_timer_interval);
}

pub fn deinit(this: *GarbageCollectionController) void {
    this.gc_timer.deinit(true);
    this.gc_repeating_timer.deinit(true);
}

pub fn scheduleGCTimer(this: *GarbageCollectionController) void {
    this.gc_timer_state = .scheduled;
    this.gc_timer.set(this, onGCTimer, 16, 0);
}

pub fn bunVM(this: *GarbageCollectionController) *VirtualMachine {
    return @alignCast(@fieldParentPtr("gc_controller", this));
}

pub fn onGCTimer(timer: *uws.Timer) callconv(.c) void {
    var this = timer.as(*GarbageCollectionController);
    if (this.disabled) return;
    this.gc_timer_state = .run_on_next_tick;
}

// We want to always run GC once in awhile
// But if you have a long-running instance of Bun, you don't want the
// program constantly using CPU doing GC for no reason
//
// So we have two settings for this GC timer:
//
//    - Fast: GC runs every 1 second
//    - Slow: GC runs every 30 seconds
//
// When the heap size is increasing, we always switch to fast mode
// When the heap size has been the same or less for 30 seconds, we switch to slow mode
pub fn updateGCRepeatTimer(this: *GarbageCollectionController, comptime setting: @Type(.enum_literal)) void {
    if (setting == .fast) {
        this.#idle_full_gcs_fired = 0;
        this.heap_size_didnt_change_for_repeating_timer_ticks_count = 0;
        if (this.gc_repeating_timer_fast) return;
        this.gc_repeating_timer_fast = true;
        this.gc_repeating_timer.set(this, onGCRepeatingTimer, this.gc_timer_interval, this.gc_timer_interval);
    } else if (setting == .slow and this.gc_repeating_timer_fast) {
        this.gc_repeating_timer_fast = false;
        this.gc_repeating_timer.set(this, onGCRepeatingTimer, 30_000, 30_000);
        this.heap_size_didnt_change_for_repeating_timer_ticks_count = 0;
    }
}

pub fn onGCRepeatingTimer(timer: *uws.Timer) callconv(.c) void {
    var this = timer.as(*GarbageCollectionController);
    if (this.disabled) return;
    const prev_heap_size = this.gc_last_heap_size_on_repeating_timer;
    var vm = this.bunVM().jsc_vm;
    const current = vm.blockBytesAllocated();
    this.gc_last_heap_size_on_repeating_timer = current;
    this.gc_last_heap_size = current;

    // Reduction mode: previous tick fired collectAsyncFull(); decide whether
    // to fire one more or converge. V8 MemoryReducer caps at 2 majors per idle.
    if (this.#idle_full_gcs_fired > 0) {
        if (current > prev_heap_size) {
            vm.collectAsync();
            this.updateGCRepeatTimer(.fast);
        } else if (prev_heap_size - current > (1 << 20) and this.#idle_full_gcs_fired < 2) {
            this.#idle_full_gcs_fired += 1;
            vm.collectAsyncFull();
        } else {
            this.#idle_full_gcs_fired = 0;
            this.heap_size_didnt_change_for_repeating_timer_ticks_count = 0;
            this.updateGCRepeatTimer(.slow);
        }
        return;
    }

    if (prev_heap_size == current) {
        this.heap_size_didnt_change_for_repeating_timer_ticks_count +|= 1;
        if (this.gc_repeating_timer_fast and this.heap_size_didnt_change_for_repeating_timer_ticks_count >= 30) {
            // 30 stable fast ticks of Eden GCs. collectAsync() never escalates
            // to Full here because Heap::updateAllocationLimits ratchets
            // m_maxHeapSize on every Eden, so the 1/3 promotion ratio decays
            // instead of crossing. Fire an explicit Full so old-gen + age-based
            // CodeBlock jettison run before we go to the 30s interval.
            this.#idle_full_gcs_fired = 1;
            vm.collectAsyncFull();
            return;
        }
    } else {
        this.updateGCRepeatTimer(.fast);
    }

    vm.collectAsync();
}

pub fn processGCTimer(this: *GarbageCollectionController) void {
    if (this.disabled) return;
    var vm = this.bunVM().jsc_vm;
    this.processGCTimerWithHeapSize(vm, vm.blockBytesAllocated());
}

fn processGCTimerWithHeapSize(this: *GarbageCollectionController, vm: *jsc.VM, this_heap_size: usize) void {
    const prev = this.gc_last_heap_size;

    switch (this.gc_timer_state) {
        .run_on_next_tick => {
            // When memory usage is not stable, run the GC more.
            if (this_heap_size != prev) {
                this.scheduleGCTimer();
                this.updateGCRepeatTimer(.fast);
            } else {
                this.gc_timer_state = .pending;
            }
            vm.collectAsync();
            this.gc_last_heap_size = this_heap_size;
        },
        .pending => {
            if (this_heap_size != prev) {
                this.updateGCRepeatTimer(.fast);

                if (this_heap_size > prev * 2) {
                    this.performGC();
                } else {
                    this.scheduleGCTimer();
                }
            }
        },
        .scheduled => {
            if (this_heap_size > prev * 2) {
                this.updateGCRepeatTimer(.fast);
                this.performGC();
            }
        },
    }
}

pub fn performGC(this: *GarbageCollectionController) void {
    if (this.disabled) return;
    var vm = this.bunVM().jsc_vm;
    vm.collectAsync();
    this.gc_last_heap_size = vm.blockBytesAllocated();
}

pub const GCTimerState = enum {
    pending,
    scheduled,
    run_on_next_tick,
};

const std = @import("std");

const bun = @import("bun");
const Environment = bun.Environment;
const uws = bun.uws;

const jsc = bun.jsc;
const VirtualMachine = jsc.VirtualMachine;

File: src/jsc/GarbageCollectionController.rs
//! Idle GC timer: JSC's own `GCActivityCallback` (via `WTFTimer`) paces eden/full against allocation rate; this adds a 1 s / 30 s idle `collect_async()` so a process that stops allocating still releases memory, and once the heap has been quiet for `BUN_IDLE_GC_SECONDS` (default "10,110,480": 10 s, 2 min and 10 min of quiet; 0 = off) full collections so JSC can age out code that no longer runs, the last of which also drops the bytecode JSC can decode again. Knobs: `BUN_GC_TIMER_INTERVAL` (ms), `BUN_GC_TIMER_DISABLE`. One per JS thread, not thread-safe.

use core::cell::Cell;
use core::ffi::c_int;

use bun_core::{Timespec, TimespecMockMode, env_var};
use bun_event_loop::EventLoopTimer::{EventLoopTimer, State as TimerState, Tag as TimerTag};
use bun_uws as uws;

use crate::JsCell;
use crate::virtual_machine::VirtualMachine;

const SLOW_REPEAT_INTERVAL_MS: i32 = 30_000;

pub struct GarbageCollectionController {
    pub gc_repeating_timer: JsCell<EventLoopTimer>,
    /// Written by every `perform_gc` caller, so the fast/slow comparison sees the last such call, not strictly the last fire; external callers are one-shot so worst case is one extra 30 s slow interval.
    pub(crate) gc_last_heap_size: Cell<usize>,
    pub(crate) heap_size_didnt_change_for_repeating_timer_ticks_count: Cell<u8>,
    pub(crate) gc_timer_interval: Cell<i32>,
    pub(crate) gc_repeating_timer_fast: Cell<bool>,
    pub(crate) disabled: Cell<bool>,
    /// Idle full collections: cumulative quiet thresholds (ms; empty = off) parsed from `BUN_IDLE_GC_SECONDS`, and the
    /// nominal time (from tick intervals) since the JS heap last grew.
    idle_gc_at_ms: Cell<[u32; 3]>,
    idle_quiet_ms: Cell<u32>,
}

bun_event_loop::impl_timer_owner!(
    GarbageCollectionController;
    from_gc_repeating_timer_ptr => gc_repeating_timer,
);

impl Default for GarbageCollectionController {
    fn default() -> Self {
        Self {
            gc_repeating_timer: JsCell::new(EventLoopTimer::init_paused(TimerTag::GcRepeating)),
            gc_last_heap_size: Cell::new(0),
            heap_size_didnt_change_for_repeating_timer_ticks_count: Cell::new(0),
            gc_timer_interval: Cell::new(0),
            gc_repeating_timer_fast: Cell::new(true),
            disabled: Cell::new(false),
            idle_gc_at_ms: Cell::new([0; 3]),
            idle_quiet_ms: Cell::new(0),
        }
    }
}

impl GarbageCollectionController {
    /// Remove `t` from the heap if linked, set its deadline to `now + ms`, and
    /// insert. JS-thread only. Real time, not the mocked clock: GC pacing is
    /// Bun's, not the test's.
    fn arm(vm: *mut VirtualMachine, t: *mut EventLoopTimer, ms: i32) {
        // SAFETY: `t` is the embedded node of the per-VM controller,
        // address-stable for the VM lifetime; JS-thread only.
        unsafe {
            if (*t).state == TimerState::ACTIVE {
                VirtualMachine::timer_remove(vm, t);
            }
            (*t).next = Timespec::now(TimespecMockMode::ForceRealTime).add_ms(i64::from(ms));
            VirtualMachine::timer_insert(vm, t);
        }
    }

    #[inline]
    fn repeat_interval(&self) -> i32 {
        if self.gc_repeating_timer_fast.get() {
            self.gc_timer_interval.get()
        } else {
            SLOW_REPEAT_INTERVAL_MS
        }
    }

    pub(crate) fn init(&self, vm: &mut VirtualMachine) {
        // SAFETY: uws::Loop::get() returns the live process-global loop.
        let actual = unsafe { &mut *uws::Loop::get() };
        actual.internal_loop_data.jsc_vm = vm.jsc_vm.cast();

        self.gc_timer_interval.set(
            env_var::BUN_GC_TIMER_INTERVAL::get()
                .filter(|&v| v > 0)
                .unwrap_or(1000)
                .min(i32::MAX as u64) as i32,
        );

        if let Some(runs) = env_var::BUN_GC_RUNS_UNTIL_SKIP_RELEASE_ACCESS::get() {
            crate::virtual_machine::Bun__defaultRemainingRunsUntilSkipReleaseAccess.store(
                runs.min(c_int::MAX as u64) as c_int,
                core::sync::atomic::Ordering::Relaxed,
            );
        }

        self.disabled
            .set(env_var::BUN_GC_TIMER_DISABLE::get().unwrap_or(false));

        // "a,b,c": seconds of quiet before the first idle full collection, then between consecutive ones (at least a
        // CodeBlock-aging lease apart so each can expire what has not run since the previous); "0"/"" = off. Every JS
        // thread, for its own heap: a Worker that has finished a burst gives its garbage back too.
        // The second at 2 min: what a program does next compiles again what that collection aged out.
        let spec = env_var::BUN_IDLE_GC_SECONDS::get().unwrap_or(b"10,110,480");
        let mut at = [0u32; 3];
        let mut sum = 0u32;
        for (slot, part) in at.iter_mut().zip(bun_core::strings::split(spec, b",")) {
            let secs = bun_core::fmt::parse_int::<u32>(bun_core::strings::trim(part, b" "), 10)
                .unwrap_or(0);
            if secs == 0 {
                break;
            }
            sum = sum.saturating_add(secs.min(3600) * 1000);
            *slot = sum;
        }
        self.idle_gc_at_ms.set(at);
    }

    /// Decides whether this tick's collection should be a full one. After the first `BUN_IDLE_GC_SECONDS` entry
    /// of ticks in which the heap did not grow, the tick's collection is made Full (it collects what the
    /// last burst left and lets JSC snapshot which code is still running), and again after each further entry of quiet:
    /// JSC drops code that has not run since the previous one, and each round makes a little more releasable (code whose
    /// last owner died in that collection, pages it emptied). Before the last one JSC also lets go of what it can get back
    /// cheaply (`shrink_footprint_now`); if it cannot right now, this tick's quiet is not counted and the next one tries
    /// again. Returns (full, ms until the next such tick is due).
    fn idle_tick(&self, vm: &VirtualMachine, grew: bool, interval_ms: i32) -> (bool, Option<u32>) {
        let dues = self.idle_gc_at_ms.get();
        if dues[0] == 0 || vm.is_inspector_enabled() {
            return (false, None);
        }
        if grew {
            self.idle_quiet_ms.set(0);
            return (false, None);
        }
        let before = self.idle_quiet_ms.get();
        let quiet = before.saturating_add(interval_ms.max(0) as u32);
        self.idle_quiet_ms.set(quiet);
        let dues = dues.into_iter().filter(|&due| due != 0);
        let crossed = |due: u32| before < due && quiet >= due;
        let mut full = dues.clone().any(crossed);
        let next = dues.clone().find(|&due| quiet < due);
        if full && next.is_none() && !vm.jsc_vm().shrink_footprint_now() {
            self.idle_quiet_ms.set(before);
            full = false;
        }
        (full, next.map(|due| due - quiet))
    }

    /// Idempotent. Must run before JSC teardown: `~RunLoop::Timer` frees the
    /// `WTFTimer` nodes sharing the heap, so an unlink afterwards walks freed
    /// siblings.
    pub(crate) fn deinit(&self) {
        self.disabled.set(true);
        let Some(vm) = VirtualMachine::get_or_null() else {
            return;
        };
        let t = self.gc_repeating_timer.as_ptr();
        // SAFETY: JS-thread; node is linked iff state == ACTIVE.
        unsafe {
            if (*t).state == TimerState::ACTIVE {
                VirtualMachine::timer_remove(vm, t);
            }
        }
    }

    /// Arms the idle timer on first call; kept at the event-loop call sites so the first deadline is in the poll that follows.
    #[inline]
    pub(crate) fn process_gc_timer(&self) {
        if self.disabled.get() || self.gc_repeating_timer.get().state != TimerState::PENDING {
            return;
        }
        let interval = self.repeat_interval();
        Self::arm(
            VirtualMachine::get_mut_ptr(),
            // whole-struct provenance: from_field_ptr recovers the container on fire
            core::ptr::addr_of!(self.gc_repeating_timer)
                .cast::<bun_event_loop::EventLoopTimer::EventLoopTimer>()
                .cast_mut(),
            interval,
        );
    }

    pub(crate) fn perform_gc(&self, idle_full: bool) {
        if self.disabled.get() {
            return;
        }
        let vm = VirtualMachine::get().jsc_vm();
        if idle_full {
            vm.collect_async_idle();
        } else {
            vm.collect_async(false);
        }
        self.gc_last_heap_size.set(vm.block_bytes_allocated());
    }

    /// `Tag::GcRepeating` fire body: `BUN_GC_TIMER_INTERVAL` (default 1 s) in fast mode, 30 s in slow mode; drops to slow after 30 fires without heap growth beyond the chatter slack, back to fast when it grows.
    ///
    /// # Safety
    /// `this` is the live per-VM controller; `vm` is the per-thread VM.
    pub unsafe fn on_gc_repeating_timer(this: *mut Self, vm: *mut VirtualMachine) {
        // SAFETY: per fn contract.
        let this = unsafe { &*this };
        this.gc_repeating_timer
            .with_mut(|t| t.state = TimerState::FIRED);
        if this.disabled.get() {
            return;
        }
        // Timer chatter in a parked app churns a few blocks per tick, a few MB per 30 s tick with nothing collecting in
        // between (a TUI at its prompt: 70-200 KB/s); real work grows the heap by far more (one request: 15 MB and up).
        const IDLE_GROWTH_SLACK_MIN: usize = 2 * 1024 * 1024;
        const IDLE_GROWTH_SLACK_PER_MS: usize = 256; // 256 KB/s: 7.5 MB over the 30 s tick
        let interval = this.repeat_interval();
        let idle_growth_slack =
            IDLE_GROWTH_SLACK_MIN.max(interval.max(0) as usize * IDLE_GROWTH_SLACK_PER_MS);
        let prev_heap_size = this.gc_last_heap_size.get();
        // SAFETY: per fn contract.
        let vm_ref = unsafe { &*vm };
        let grew = vm_ref.jsc_vm().block_bytes_allocated() > prev_heap_size + idle_growth_slack;
        let (full, idle_gc_due_in) = this.idle_tick(vm_ref, grew, interval);
        this.perform_gc(full);
        // Growth is activity; a shrinking heap is a collection (possibly the one requested above) doing its job. Where the
        // event loop cannot let an idle collection finish while parked (Windows: Bun__JSC_onBeforeWait), it proceeds at
        // this timer's ticks instead, fast ones for the next 30.
        let needs_ticks_to_finish = full && cfg!(windows);
        if !grew && !needs_ticks_to_finish {
            let ticks = this
                .heap_size_didnt_change_for_repeating_timer_ticks_count
                .get()
                .saturating_add(1);
            this.heap_size_didnt_change_for_repeating_timer_ticks_count
                .set(ticks);
            if ticks >= 30 {
                this.gc_repeating_timer_fast.set(false);
            }
        } else {
            this.heap_size_didnt_change_for_repeating_timer_ticks_count
                .set(0);
            this.gc_repeating_timer_fast.set(true);
        }
        let interval = match idle_gc_due_in {
            Some(ms) => this.repeat_interval().min(ms.max(1000) as i32),
            None => this.repeat_interval(),
        };
        Self::arm(
            vm,
            // whole-struct provenance: from_field_ptr recovers the container on fire
            core::ptr::addr_of!(this.gc_repeating_timer)
                .cast::<bun_event_loop::EventLoopTimer::EventLoopTimer>()
                .cast_mut(),
            interval,
        );
    }
}

impl Drop for GarbageCollectionController {
    fn drop(&mut self) {
        self.deinit();
    }
}
